traitsui/tests/editors/test_boolean_editor.py

```diff
@@ -43,6 +43,19 @@ def test_init_dispose(self):
         # Test init and dispose of the editor.
         view = View(Item("true_or_false", editor=BooleanEditor()))
         obj = BoolModel()
-        with reraise_exceptions(), \
-                create_ui(obj, dict(view=view)):
+
+        tester = UITester()
+        with tester.create_ui(obj, dict(view=view)):
             pass
+
+    def test_click_boolean(self):
+        view = View(Item("true_or_false", editor=BooleanEditor()))
+        obj = BoolModel()
+
+        tester = UITester()
+        with tester.create_ui(obj, dict(view=view)) as ui:
+            # sanity check
+            self.assertEqual(obj.true_or_false, False)
+            checkbox = tester.find_by_name(ui, "true_or_false")
+            checkbox.perform(MouseClick())
+            self.assertEqual(obj.true_or_false, True)
```

Comment:
There are a number of recommended testing practices: [test pyramid](https://martinfowler.com/bliki/TestPyramid.html), [testing trophy](https://kentcdodds.com/blog/write-tests). But broadly speaking, we like tests that are
(1) fast and cheap to run, because then developers will be more likely to run them frequently, get test feedbacks often and catch errors sooner.
(2) small and focused so that its failure help pinning down the source of bug more quickly. This also allows tests targeting different behaviour to be run independently of each other. If all are lumped into a single test, the first failure stops the rest from being run so you don't get to see the whole picture early.

The test in the demo, if run on its own, is not slow, but it does lump a lot of things together. Furthermore, it does not live with the `traitsui.tests.editors` where developers find editor specific tests. Indeed, the tests in the demo domain serve different purposes: it protects the demo script and it serves to demonstrate writing tests. The demo test changes if the demo changes, which could be caused by educational reasons. The demo could well have lived in a separate repository outside of TraitsUI. With that, we should not rely on the demo tests to protect the editor's behaviour.

Imagine a developer changing the code for BooleanEditor, they will likely notice `test_boolean_editor` to be the tests for BooleanEditor, not the one in the demo. The test in the demo only gets picked up when they run the integration test suite. The integration test suite is a lot slower that one may not run as often.

Suggestion: Could you move the the test logic you have in the demo here, and organize them such that different test objectives can be run independently (e.g. separating simple editor and custom editor, separating trait change -> GUI change and GUI change -> trait change)? Then the demo test can also be simplified.

(Unrelated comment: When we have to add tests as an afterthought to code that doesn't have tests, we tend to have to go backwards and have to write bigger characteristics tests first to preserve behaviour before we can refactor the code to make it easier to test in small bits and pieces.)

Reply:
Hi @kitchoi, thank you for the review!  I added more simple unit tests to `test_boolean_editor` as you suggested.  I'm not exactly sure what should be removed from the demo tests to simplify them now like you mentioned.  
Also, as we discussed I added an `IsChecked` class to `query` to better be able to test trait change -> GUI change.

Reply:
>  I'm not exactly sure what should be removed from the demo tests to simplify them now like you mentioned.

I was just wondering if you'd have preferred the demo to be different if the test logic is moved here, because after all there are some duplications. If what is in the demo test now is what you'd have written regardless of what is in `test_boolean_editor`, then it is totally fine.

> Also, as we discussed I added an IsChecked class to query to better be able to test trait change -> GUI change.

Thanks!